File: app/src/main/java/mono/hg/fragments/BackupRestoreFragment.java
package mono.hg.fragments;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;

import java.io.File;
import java.io.FileFilter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import mono.hg.R;
import mono.hg.SettingsActivity;
import mono.hg.adapters.FileFolderAdapter;
import mono.hg.models.FileFolder;
import mono.hg.utils.BackupRestoreUtils;
import mono.hg.wrappers.BackHandledFragment;

public class BackupRestoreFragment extends BackHandledFragment {
    private ArrayList<FileFolder> fileFoldersList = new ArrayList<>();
    private FileFolderAdapter fileFolderAdapter;
    private File currentPath;
    private Boolean isInRestore = false;
    private EditText backupNameField;

    @Override public View onCreateView(@NonNull LayoutInflater inflater,
            ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_backup_restore, container, false);
    }

    @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (getArguments() != null) {
            isInRestore = getArguments().getBoolean("isRestore", false);
        }

        super.onCreate(savedInstanceState);

        /*
         * Show our own toolbar if we are running below Lollipop.
         * This is needed because the preference library does not supply toolbars
         * for fragments that it isn't managing.
         */
        ActionBar actionBar = ((SettingsActivity) requireActivity()).getSupportActionBar();
        if (actionBar != null) {
            if (isInRestore) {
                actionBar.setTitle(R.string.pref_header_restore);
            } else {
                actionBar.setTitle(R.string.pref_header_backup);
            }
        }

        setHasOptionsMenu(true);

        fileFolderAdapter = new FileFolderAdapter(fileFoldersList, requireActivity());
        FrameLayout fileInputContainer = requireActivity().findViewById(R.id.file_input_container);
        ListView fileFolders = requireActivity().findViewById(R.id.files_list);
        backupNameField = requireActivity().findViewById(R.id.file_input_entry);

        fileFolders.setAdapter(fileFolderAdapter);

        // If we are called to restore, then hide the input field.
        if (isInRestore) {
            fileInputContainer.setVisibility(View.GONE);
        }

        // Check for storage permission if we're in Marshmallow and up.
        currentPath = Environment.getExternalStorageDirectory();
        traverseStorage(currentPath);
        fileFolders.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                // Open and traverse to directory.
                if (fileFoldersList.get(position).isFolder()) {
                    currentPath = new File(
                            currentPath + File.separator + fileFoldersList.get(position).getName());
                    traverseStorage(currentPath);
                } else {
                    // Restore backup when clicking a file, but only do so in restore mode.
                    String possibleBackup = "file://" + currentPath + File.separator + fileFoldersList.get(position).getName();
                    if (isInRestore && fileFoldersList.get(position).getName().indexOf('.') > 0) {
                        new BackupRestoreUtils.restoreBackupTask((SettingsActivity) requireActivity(), possibleBackup).execute();
                    }
                }
            }
        });
    }

    @Override public boolean onBackPressed() {
        if (!currentPath.getPath().equals(Environment.getExternalStorageDirectory().getPath())) {
            currentPath = new File(currentPath.getParent());
            traverseStorage(currentPath);
            return true;
        } else {
            return false;
        }
    }

    @Override public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        menu.clear();
        menu.add(0, 1, 100, getString(R.string.action_backup));
        menu.getItem(0).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
        menu.getItem(0).setVisible(!isInRestore);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case android.R.id.home:
                requireActivity().onBackPressed();
                return true;
            case 1:
                // Send our backup signal!
                if (!backupNameField.getText().toString().equals("")) {
                    final String backupPath = "file://" + currentPath + File.separator + backupNameField.getText().toString() + ".xml";
                    final File backupName = new File(backupPath);
                    if (backupName.exists() && backupName.isFile()) {
                        final AlertDialog.Builder overwriteDialog = new AlertDialog.Builder(
                                requireActivity());
                        overwriteDialog.setTitle(getString(R.string.pref_header_backup));
                        overwriteDialog.setMessage(getString(R.string.backup_exist));
                        overwriteDialog.setNegativeButton(getString(R.string.backup_exist_cancel),
                                null);
                        overwriteDialog.setPositiveButton(
                                getString(R.string.backup_exist_overwrite),
                                new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        BackupRestoreUtils.saveBackup(requireActivity(), backupPath);
                                        traverseStorage(currentPath);
                                    }
                                });
                        overwriteDialog.show();
                    } else {
                        BackupRestoreUtils.saveBackup(requireActivity(), backupPath);
                        traverseStorage(this.currentPath);
                    }
                } else {
                    Toast.makeText(requireActivity(), R.string.backup_empty, Toast.LENGTH_SHORT)
                         .show();
                }
                return true;
            default:
                // Do nothing.
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    // Open a directory and refresh fileFoldersList.
    private void traverseStorage(File path) {
        fileFoldersList.clear();
        fileFolderAdapter.notifyDataSetInvalidated();
        File[] contents;
        if (isInRestore) {
            contents = path.listFiles(new FileFilter() {
                public boolean accept(File dir) {
                    return dir.getName().toLowerCase().endsWith(".xml")
                            || dir.isDirectory() && !dir.isFile();
                }
            });
        } else {
            contents = path.listFiles();
        }

        if (contents != null && contents.length > 0) {
            for (File availableContents : contents) {
                // Don't show hidden (.dot) files/folders.
                if (!availableContents.isHidden()) {
                    fileFoldersList.add(new FileFolder(availableContents.getName(),
                            availableContents.isDirectory()));
                }
            }
            fileFolderAdapter.notifyDataSetChanged();
        }

        Collections.sort(fileFoldersList, new Comparator<FileFolder>() {
            @Override
            public int compare(FileFolder f1, FileFolder f2) {
                if (f1.isFolder() && !f2.isFolder()) {
                    return -1;
                } else if (!f1.isFolder() && f2.isFolder()) {
                    return 1;
                } else {
                    return f1.getName().compareToIgnoreCase(f2.getName());
                }
            }
        });
    }
}

File: app/src/main/java/mono/hg/fragments/BackupRestoreFragment.kt
package mono.hg.fragments

import android.app.AlertDialog
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.EditText
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import mono.hg.R
import mono.hg.SettingsActivity
import mono.hg.adapters.FileFolderAdapter
import mono.hg.databinding.FragmentBackupRestoreBinding
import mono.hg.models.FileFolder
import mono.hg.utils.BackupRestoreUtils
import mono.hg.utils.compatHide
import mono.hg.utils.compatShow
import java.io.File
import java.util.*
import kotlin.Comparator
import kotlin.collections.ArrayList

/**
 * A Fragment that displays both backup and restore options.
 */
class BackupRestoreFragment : Fragment() {
    private var binding: FragmentBackupRestoreBinding? = null
    private val fileFoldersList = ArrayList<FileFolder>()
    private var fileFolderAdapter: FileFolderAdapter? = null
    private var currentPath: File? = null
    private var isInRestore = false
    private lateinit var backupNameField: EditText

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        binding = FragmentBackupRestoreBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // We are using the back press to traverse back
        // in the path, so don't let the activity consume the event
        // unless we're already at root path.
        activity?.onBackPressedDispatcher?.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                @Suppress("DEPRECATION")
                if (currentPath?.path != Environment.getExternalStorageDirectory().path) {
                    currentPath?.parent?.apply {
                        currentPath = File(this)
                        traverseStorage(currentPath)
                    }
                } else {
                    isEnabled = false
                    activity?.onBackPressed()
                }
            }
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        isInRestore = arguments?.getBoolean("isRestore", false) ?: false

        super.onCreate(savedInstanceState)

        /*
         * Show our own toolbar if we are running below Lollipop.
         * This is needed because the preference library does not supply toolbars
         * for fragments that it isn't managing.
         */
        (requireActivity() as SettingsActivity).supportActionBar?.apply {
            if (isInRestore) {
                this.setTitle(R.string.pref_header_restore)
            } else {
                this.setTitle(R.string.pref_header_backup)
            }
        }

        setHasOptionsMenu(true)
        fileFolderAdapter = FileFolderAdapter(fileFoldersList, requireContext())
        val fileFolders = binding !!.filesList
        backupNameField = binding !!.fileInputEntry
        fileFolders.adapter = fileFolderAdapter

        // If we are called to restore, then hide the input field.
        if (isInRestore) {
            binding !!.fileInputContainer.visibility = View.GONE
        }

        // This fragment is only used for API levels that does not specifically
        // need scoped storage. These API levels will likely not require
        // any change, and the deprecation warning is pointless.
        @Suppress("DEPRECATION")
        currentPath = Environment.getExternalStorageDirectory()
        traverseStorage(currentPath)
        fileFolders.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            // Open and traverse to directory.
            if (fileFoldersList[position].isFolder) {
                currentPath = File(
                    currentPath.toString() + File.separator + fileFoldersList[position].name
                )
                traverseStorage(currentPath)
            } else {
                // Restore backup when clicking a file, but only do so in restore mode.
                val possibleBackup =
                    "file://" + currentPath + File.separator + fileFoldersList[position].name
                if (isInRestore && fileFoldersList[position].name.indexOf('.') > 0) {
                    lifecycleScope.launch {
                        BackupRestoreUtils.restoreBackup(
                            requireActivity() as SettingsActivity,
                            possibleBackup
                        )
                    }
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        with(menu) {
            clear()
            add(0, 1, 100, getString(R.string.action_backup))
            getItem(0).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
            getItem(0).isVisible = ! isInRestore
            super.onCreateOptionsMenu(this, inflater)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.home -> {
                requireActivity().onBackPressed()
                true
            }
            1 -> {
                // Send our backup signal!
                if (backupNameField.text.toString() != "") {
                    val backupPath =
                        "file://" + currentPath + File.separator + backupNameField.text.toString() + ".xml"
                    val backupName = File(backupPath)
                    if (backupName.exists() && backupName.isFile) {
                        with(AlertDialog.Builder(requireActivity())) {
                            setTitle(getString(R.string.pref_header_backup))
                            setMessage(getString(R.string.backup_exist))
                            setNegativeButton(getString(R.string.dialog_cancel), null)
                            setPositiveButton(getString(R.string.backup_exist_overwrite)) { _, _ ->
                                BackupRestoreUtils.saveBackup(requireActivity(), backupPath)
                                traverseStorage(currentPath)
                            }
                            show()
                        }
                    } else {
                        BackupRestoreUtils.saveBackup(requireActivity(), backupPath)
                        traverseStorage(currentPath)
                    }
                } else {
                    Toast.makeText(requireActivity(), R.string.backup_empty, Toast.LENGTH_SHORT)
                        .show()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    // Open a directory and refresh fileFoldersList.
    private fun traverseStorage(path: File?) {
        fileFoldersList.clear()
        fileFolderAdapter?.notifyDataSetInvalidated()
        val contents: Array<File>?
        contents = if (isInRestore) {
            path?.listFiles { dir ->
                (dir.name.toLowerCase(Locale.getDefault())
                    .endsWith(".xml") || dir.isDirectory && ! dir.isFile)
            }
        } else {
            path?.listFiles()
        }
        if (contents != null && contents.isNotEmpty()) {
            lifecycleScope.launch {
                withContext(Dispatchers.Default) {
                    contents.filter { ! it.isHidden }
                        .forEach { fileFoldersList.add(FileFolder(it.name, it.isDirectory)) }
                }

                fileFoldersList.sortWith(Comparator { f1, f2 ->
                    if (f1.isFolder && ! f2.isFolder) {
                        - 1
                    } else if (! f1.isFolder && f2.isFolder) {
                        1
                    } else {
                        f1.name.compareTo(f2.name, ignoreCase = true)
                    }
                })
                (requireActivity() as SettingsActivity).progressBar.compatShow()
                fileFolderAdapter?.notifyDataSetChanged()
                (requireActivity() as SettingsActivity).progressBar.compatHide()
            }
        }
    }
}